### ISSUE TITLE
TagHelper registration

### ISSUE THREAD

**Issue by dazinator** [aspnet/Home#1684]
I have a TagHelper in some assembly. 
When my application starts, I load the assembly with the taghelper into the app domain, and register it with mvc's application parts system.

```
var assy = Assembly.LoadFile("C:\\SomePath\SomeAssembly.dll");
mvcBuilder.AddApplicationPart(assy);
```

My application does not reference this assembly at compile time, it loads it dynamically.

I then have a razor view in my application that uses the tag. 

The goal I'd like to achieve is that when the TagHelper is loaded (i.e the assembly is present), I want the tag to work, but if the assembly is not present, the tag won't work, and that's fine - i'll see the original markup left in my html output.

The problem is that I notice that the TagHelper doesn't work at all, unless I also add this line to `_ViewImports.cshtml`

```
@addTagHelper "*, SomeAssembly"
```

However I can only add that line, when the assembly is referenced from my project, otherwise I get this error on application startup:

```

An error occurred during the compilation of a resource required to process this request. Please review the following specific error details and modify your source code appropriately.

/Views/_ViewImports.cshtml

Cannot resolve TagHelper containing assembly 'SomeAssembly'. Error: Could not load file or assembly 'SomeAssembly' or one of its dependencies. The system cannot find the file specified.
@addTagHelper "*, SomeAssembly"

```

What do I have to do in addition to registering the the assembly as an application part, in order for it to work in razor views? I'm guessing it also needs to be dynamically included in the razor compilation process somehow but I am not sure how.

Perhaps there is a better way to achieve what I want? What am I missing..

**Comment by dazinator** [comment]
Ok I experimented and found something interesting.

I embedded a `Views.Shared._ViewImports.cshtml` file in the tag helper assembly. When I register it with the `ApplicationParts` system, I also set up an `EmbeddedFileProvider` and include it in the  `IHostingEnvironment` s `ContentRootFileProvider` (which is a composite file provider).

Essentially this now means there are two` _ViewImports.cshtml` files visible to the application, one from the physical file system, and one that is embedded in my taghelper assembly:

```
/Views/_ViewImports.cshtml
/Views/Shared/_ViewImports.cshtml
```

This now works the way i want.. When the assembly is present the tag works. When it isn't the tag is just left in my html output.

However this doesn't completely solve the problem because it's not scalable. If I do this with another assembly, there is no where else in the folder hierarchy to include yet another `ViewImports.cshtml` file... I can't include two files in the same directory.

**Comment by dazinator** [comment]
I'm starting to wonder if the only way to do this, is to dynamically create a `_ViewImports.cshtml` file, based on the assemblies i'm loading. Seems horrid so i'm very hesitant.

**Comment by pranavkm** [comment]
cc @NTaylorMullen.

**Comment by dazinator** [comment]
Tried also adding this line but it made no difference.. In fact i'm not actually sure what this does:

```
   mvcBuilder.AddTagHelpersAsServices();

```

**Comment by dazinator** [comment]
I think what i might need to do is implement and register my own `ITagHelperDescriptorResolver` and always resolve any tag helper that lives in my dynamic assemblies. If there is an easier way please let me know!

**Comment by NTaylorMullen** [comment]
If you're trying to automatically include TagHelpers from an assembly at runtime my recommendation is to 

1. Ensure the assembly is loadable (sounds like you've done this)
2. Modify the MVC Razor host's [default inherited chunks](https://github.com/aspnet/Mvc/blob/659cdc336484039f4c2e619ddbf1cc4f09459523/src/Microsoft.AspNetCore.Mvc.Razor.Host/MvcRazorHost.cs#L195) to include an additional addTagHelper directive for your assembly of choice.

**Comment by dazinator** [comment]
@NTaylorMullen - thanks. Not going to pretend I know what a chunk is in the context of razor but, what's the best way to go about getting access to the `MvcRazorHost` so that I can modify this property?

**Comment by NTaylorMullen** [comment]
You can override the `IMvcRazorHost` interface in DI with your own implementation. MvcRazorHost is the default host that MVC registers.

**Comment by rdefreitas** [comment]
Hmmm, I think the way to go is to do the ITagHelperDescriptorResolver implementation.  I have need for something similar on my end and I'm thinking of following the same approach and having it automatically resolve anything that implements ITagHelper from all currently loaded assemblies.

The only downside to that is concerns over collision and which tag helper from which takes precedence.

**Comment by dazinator** [comment]
@rdefreitas yeah I am thinking on along the same lines. Can you share if you get anything working?

What i don't understand at the moment, is if you use:

```
  mvcBuilder.AddTagHelpersAsServices();
```

This will then allow you to have an `addTagHelper` directive to an unknown assembly (at compile time) and it wont blow up. However when the specified assembly is loaded at runtime, and registered with the application parts manager, the tag helper still doesn't work. Shouldn't the default `ITagHelperDescriptorResolver` now resolve it assuming its seeing all assemblies registered with the parts manager?

> The only downside to that is concerns over collision and which tag helper from which takes precedence.

Yeah, I think this is where @NTaylorMullen's suggestion of having a another directive might help, because you can pull in your new directives and use those to only load the tag helpers you need from your custom assemblies. My problem is that I don't understand (based on what I said above) why this isn't already working. I'll step through the code some more when I get another chance :-)

**Comment by NTaylorMullen** [comment]
@dazinator the current TagHelper system doesn't look at AssemblyParts when resolving assemblies. It just does [Assembly.Load](https://github.com/aspnet/Razor/blob/d337bacc69fe4ca39532d958712738c71f07c091/src/Microsoft.AspNetCore.Razor.Runtime/Runtime/TagHelpers/TagHelperTypeResolver.cs#L72).

**Comment by dazinator** [comment]
@NTaylorMullen oooo that's interesting!

So there's potential for me to implement a custom `ITagHelperTypeResolver` in which I can also resolve from any assemblies i have manually loaded from disk?

I'll give that a go tomorrow and fingers crossed.

Just out of curiosity is there a reason why `ITagHelperTypeResolver` isn't including application parts assemblies - seems odd that it wouldn't given that this is the primary mechanism now of hooking assemblies / types up with MVC discovery?

**Comment by dazinator** [comment]
Ok so, `mvcBuilder.AddTagHelpersAsServices()` registers a `ITagHelperTypeResolver` that uses `ApplicationParts`, but without that API call, the default `ITagHelperTypeResolver` just does an `Assembly.Load()` so doesn't work with the parts system.

@NTaylorMullen - Wouldn't it be better for MVC to just register the parts friendly services by default? Seems like there would no reason not to do this from my perhaps limited perspective..

Anyway, just being able to resolve the `TagHelper` types is not enough to get them to "automatically" be available / used on razor pages - you still need to include the directives, which is where @NTaylorMullen's original suggestion of creating a custom razor host to automatically inject them comes in.

I'm first going to exhaust the custom `ITagHelperDescriptorResolver` route to see if I can inject them that way, but may give up on that we shall see!

**Comment by dazinator** [comment]
Ok I think  I have come up with a nice solution for this.

The idea is to support globbing for the assembly name matching in the `ITagHelperTypeResolver` and then you can add a single directive to viewImports:

```
@addTagHelper "*, Gluon.Module.*"
```

It doesn't get around the need to have to manually add the directive, but atleast you only have to add 1, and when new assemblies are registered with the pars system they should just light up..

I think this feature (globbing on the assembly name) should be supported by mvc proper, so design time host can benefit from it also..

**Comment by NTaylorMullen** [comment]
Glad you found an approach that you were comfortable with. 

As for supporting globing on assembly names within MVC. I'm not sure there's enough of a main stream use-case other than to make it easier to not type out the entire TagHelper assembly name. It'd also open the door for new assemblies added to your project to automatically add their TagHelpers to your page (may not always be the intended case of adding a new reference).

In my opinion this would make a lot more sense as an external package that users can bring into their applications. Could make peoples lives a lot easier if they're looking for this functionality :smile:

**Comment by dazinator** [comment]
> It'd also open the door for new assemblies added to your project to automatically add their TagHelpers to your page (may not always be the intended case of adding a new reference).

Yeah that's kind of the intention though if you specify a glob pattern for the assembly in your `addTagHelper` directive. You wouldn't do that by default unless seeking this kind of functionality.

It's just odd because MVC already supports a glob pattern on the type name portion of the directive, this is just extending that to it's logical conclusion!

I've got this working though, so i'm happy. I might put a re-usable package up if I get some time, because I think there are some potential other nice features that could be worth implementing. One idea I have at the moment, is that I could wrap all `TagHelper`'s returned with a special implementation that defers to the underlying TagHelper when there is one resolved, or in the case that no `TagHelper`'s were resolved, then it would act as the fallback. What this fallback could do, is remove all Html elements that adhere to some convention. So if for example you are using tags like this:

<awesome-tag></awesome-tag>

and that TagHelper doesn't load, then the fallback `TagHelper` would remove all tags prefixed `<awesome-` from the document..
